### PR TITLE
Add support for Cloudflare Workers

### LINKED ISSUE #4948: Add Support for Cloudflare Workers
# This is a Feature Proposal

## Description

I'm an engineer on the [Workers](https://developers.cloudflare.com/workers/about/) team at Cloudflare. We'd like to start a conversation around adding support for deploying to Workers to serverless.

Workers is similar to Lambda, but your code runs on Cloudflare's edge around the world (155 data centers at the moment). This means you can run your application directly on what most people would think of as the CDN layer.

It's JavaScript based, but built on V8 directly, rather than Node, meaning its environment is more lightweight and more similar to that in a browser. The API to write a Worker is based on the Service Worker spec, and you can use build tools like Webpack to assemble files and dependencies into a Worker.

We're of the opinion that in the future people will choose to deploy not just serverless-ly, but 'originless'; That the default will be having your code run close to all your users, not in a single location. The storage infrastructure isn't necessarily there yet to accomplish that goal fully, but for things like API Gateways and stateless applications it's possible now.

### PR DESCRIPTION
**Pull request by pinkerton**
## What did you implement:

Closes #4948 

This PR adds templates and documentation for Cloudflare Workers as a part of the provider plugin that we're reviewing and will publish on NPM very soon.

## How did you implement it:

Added `cloudflare-workers` and `cloudflare-workers-enterprise` templates and whitelisted them.

Regarding documentation, I am not 100% sure how this is generated and how images will be hosted (I see they are mostly in an S3 bucket).

## How can we verify it:

1. Set up a Cloudflare account and enable Workers
1. `export CLOUDFLARE_AUTH_KEY=YOUR_CLOUDFLARE_GLOBAL_API_KEY` (get your Global API key from your [Cloudflare profile](https://dash.cloudflare.com/profile) page).
1. `export CLOUDFLARE_AUTH_EMAIL=you@example.com`
1. `git clone https://github.com/pinkerton/serverless.git`
1. `git clone https://github.com/cloudflare/serverless-cloudflare-workers.git`
1. `cd serverless`
1. `npm install`
1. `npm link ../serverless-cloudflare-workers`
1. `cd ..`
1. `./serverless/bin/serverless create --template cloudflare-workers --path my-project`
1. `cd my-project`
1. `npm install`
1. `npm link ../serverless-cloudflare-workers`
1. Get your account ID and put it in `serverless.yml` (get it from the URL when using the [Cloudflare dashboard](https://dash.cloudflare.com)).
1. Get your zone ID and put it in `serverless.yml` (`zoneId` from the `overview` tab after selecting the desired zone from the [Cloudflare dashboard](https://dash.cloudflare.com)). 
1. Update your zone and routes in `serverless.yml`.
1. `serverless deploy`
1. `serverless invoke -f helloWorld`

### `serverless create`

serverless-create

### `serverless deploy` & `serverless invoke`

serverless-deploy-invoke

## Todos:

- [ ] Write tests
- [x] Write documentation
- [ ] Fix linting errors
- [ ] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [ ] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by pinkerton** [comment]
@horike37 That was accidental. I must have committed it by mistake when rebasing / merging. Will revert.

**Comment by pinkerton** [comment]
Un-committed the accidental README changes.

File: docs/providers/cloudflare-workers/guide/deploying.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+ 
+For example, let's take the following example `serverless.yml` file:
+ 
+```yml
+# serverless.yml
+
+service:
+    name: hello-world
+    config:
+      accountId: CLOUDFLARE_ACCOUNT_ID
```
[the remark this answers is not in the document]
Comment (pinkerton) [reply docs/providers/cloudflare-workers/guide/deploying.md]: Interesting, that may have been left in accidentally. I'm double-checking if we can remove that.

[the remark this answers is not in the document]
Comment (nirvik) [reply docs/providers/cloudflare-workers/guide/deploying.md]: So, if the user has the `multiscript` feature enabled, the api call uses `accountId` as one of the parameters.
https://github.com/cloudflare/serverless-cloudflare-workers/blob/master/deploy/lib/multiscript.js#L27

Whereas, for `singleScript` feature, `accountId` is not required. 
https://github.com/cloudflare/serverless-cloudflare-workers/blob/master/deploy/lib/singlescript.js#L27


All in all, `accountId` is *required* in MultiScript, but *optional* in SingleScript

**Comment by pinkerton** [comment]
Thanks so much for taking a look @raeesbhatti. 

What version of the CLI tool will this ship in? It looks like maybe 1.30.4 or 1.31.0. Our docs have an old version in them as a placeholder for the minimum version that I'd like to update.

I also need to add our images to our documentation.

And how do we get our docs linked on this page https://serverless.com/framework/docs/ ?

**Comment by raeesbhatti** [comment]
@pinkerton The docs on the website will be updated automatically on new build. But I noticed that `serverless-cloudflare-workers ` plugin is not published on npm, we're planning to publish a post on Tuesday at Serverless blog regarding Cloudflare support. It would be ideal if the plugin was published before that.
@eahefnawy Can you please address the question about versioning. Thanks

**Comment by pinkerton** [comment]
Thanks @raeesbhatti. I'm working on publishing our plugin to NPM now. Running into an issue where I can publish to our org as an org-scoped package, but that would make the name `@cloudflare/serverless-cloudflare-workers`, which I'm trying to avoid. I can publish it without the scope to my own profile, however.

Will have this resolved soon.

**Comment by eahefnawy** [comment]
@pinkerton it'll likely be `v1.31.0`, since there are tons of new, awesome stuff in here! We really appreciate it!

**Comment by pinkerton** [comment]
Thanks for catching that @raeesbhatti

**Pull request merged**